zcash_client_backend/src/data_api/wallet.rs

```diff
@@ -772,104 +795,108 @@ where
+            None if !ephemeral_added => {
+                // The ephemeral addresses are added lazily to avoid extra database operations
+                // in the common case. We don't need to include them in order to be able to
+                // construct ZIP 320 transactions, because in that case the ephemeral output
+                // is represented via a "change" reference to a previous step. However, we do
+                // need them in order to create a transaction from a proposal that explicitly
+                // spends an output from an ephemeral address. This need not set `for_detection`
+                // because we only need to be able to spend outputs already detected by this
+                // wallet instance.
+                ephemeral_added = true;
+                known_addrs.extend(
+                    wallet_db
+                        .get_reserved_ephemeral_addresses(account_id, false)
+                        .map_err(Error::DataSource)?
+                        .into_iter(),
+                );
+                known_addrs.get(&addr)
+            }
+            result => result,
+        }
+        .ok_or(Error::AddressNotRecognized(addr))?
+        .clone()
+        .ok_or_else(|| Error::NoSpendingKey(addr.encode(params)))
+    };
 
+    #[cfg(feature = "transparent-inputs")]
+    let utxos_spent = {
         let mut utxos_spent: Vec<OutPoint> = vec![];
-        let mut add_transparent_input = |addr: &TransparentAddress,
+        let mut add_transparent_input = |address_metadata: &TransparentAddressMetadata,
                                          outpoint: OutPoint,
-                                         utxo: TxOut|
-         -> Result<
-            (),
-            Error<
-                <DbT as WalletRead>::Error,
-                <DbT as WalletCommitmentTrees>::Error,
-                InputsErrT,
-                FeeRuleT::Error,
-            >,
-        > {
-            let address_metadata = known_addrs
-                .get(addr)
-                .ok_or(Error::AddressNotRecognized(*addr))?
-                .clone()
-                .ok_or_else(|| Error::NoSpendingKey(addr.encode(params)))?;
-
+                                         txout: TxOut|
+         -> Result<(), ErrorT<DbT, InputsErrT, FeeRuleT>> {
             let secret_key = usk
                 .transparent()
                 .derive_secret_key(address_metadata.scope(), address_metadata.address_index())
-                .unwrap();
+                .expect("spending key derivation should not fail");
 
             utxos_spent.push(outpoint.clone());
-            builder.add_transparent_input(secret_key, outpoint, utxo)?;
+            builder.add_transparent_input(secret_key, outpoint, txout)?;
 
             Ok(())
         };
 
         for utxo in proposal_step.transparent_inputs() {
             add_transparent_input(
-                utxo.recipient_address(),
+                &metadata_from_address(*utxo.recipient_address())?,
```

Comment:
Reiterating my discomfort with `metadata_from_address` quietly mutating variables in the enclosing scope. I would vastly prefer if the mutations were explicitly statements in the loop; why is it necessary to use a closure here? It obscures the mutation operations, which are the critical bit. At very least I would like the mutations described in a comment that is local to this call.

Reply:
I take your point with the `add_*` functions because there the mutation of the metadata vectors is an observable intended side effect. For `metadata_from_address` the mutable state is just a cache (more obviously so now).